flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/FileSink.java

```diff
@@ -0,0 +1,434 @@
+	}
+
+	/**
+	 * A builder for configuring the sink for row-wise encoding formats.
+	 */
+	public static class RowFormatBuilder<IN, BucketID, T extends RowFormatBuilder<IN, BucketID, T>>
+			extends BucketsBuilder<IN, BucketID, T> {
+
+		private static final long serialVersionUID = 1L;
+
+		private final Path basePath;
+
+		private final Encoder<IN> encoder;
+
+		private BucketAssigner<IN, BucketID> bucketAssigner;
+
+		private RollingPolicy<IN, BucketID> rollingPolicy;
+
+		private FileWriterBucketFactory<IN, BucketID> bucketFactory;
+
+		private OutputFileConfig outputFileConfig;
+
+		protected RowFormatBuilder(
+				Path basePath,
+				Encoder<IN> encoder,
+				BucketAssigner<IN, BucketID> bucketAssigner) {
+			this(
+					basePath,
+					encoder,
+					bucketAssigner,
+					DefaultRollingPolicy.builder().build(),
+					new DefaultFileWriterBucketFactory<>(),
+					OutputFileConfig.builder().build());
+		}
+
+		protected RowFormatBuilder(
+				Path basePath,
+				Encoder<IN> encoder,
+				BucketAssigner<IN, BucketID> assigner,
+				RollingPolicy<IN, BucketID> policy,
+				FileWriterBucketFactory<IN, BucketID> bucketFactory,
+				OutputFileConfig outputFileConfig) {
+			this.basePath = checkNotNull(basePath);
+			this.encoder = checkNotNull(encoder);
+			this.bucketAssigner = checkNotNull(assigner);
+			this.rollingPolicy = checkNotNull(policy);
+			this.bucketFactory = checkNotNull(bucketFactory);
+			this.outputFileConfig = checkNotNull(outputFileConfig);
+		}
+
+		public T withBucketAssigner(final BucketAssigner<IN, BucketID> assigner) {
+			this.bucketAssigner = checkNotNull(assigner);
+			return self();
+		}
+
+		public T withRollingPolicy(final RollingPolicy<IN, BucketID> policy) {
+			this.rollingPolicy = checkNotNull(policy);
+			return self();
+		}
+
+		public T withOutputFileConfig(final OutputFileConfig outputFileConfig) {
+			this.outputFileConfig = outputFileConfig;
+			return self();
+		}
+
+		public RowFormatBuilder<IN, BucketID, ? extends RowFormatBuilder<IN, BucketID, ?>> withNewBucketAssignerAndPolicy(
+				BucketAssigner<IN, BucketID> assigner,
+				RollingPolicy<IN, BucketID> policy) {
+			Preconditions.checkState(
+					bucketFactory.getClass() == DefaultFileWriterBucketFactory.class,
+					"newBuilderWithBucketAssignerAndPolicy() cannot be called "
+							+ "after specifying a customized bucket factory");
+			return new RowFormatBuilder<>(
+					basePath,
+					encoder,
+					checkNotNull(assigner),
+					checkNotNull(policy),
+					bucketFactory,
+					outputFileConfig);
+		}
+
+		@VisibleForTesting
+		T withBucketFactory(final FileWriterBucketFactory<IN, BucketID> factory) {
+			this.bucketFactory = Preconditions.checkNotNull(factory);
+			return self();
+		}
+
+		/** Creates the actual sink. */
+		public FileSink<IN, BucketID> build() {
+			return new FileSink<>(this);
+		}
+
+		@Override
+		public FileWriter<IN, BucketID> createWriter(InitContext context) throws IOException {
```

Comment:
This can become `package-private`.

Reply:
It seems we could not change the modifier since the method is inherited from an interface~?

Reply:
The interface is the `BucketsBuilder` right? Can't we change that to `package-private` as well? It is a newly introduced interface.

Reply:
Yes, indeed, I changed the modifier to `package-private`.